src/Dictionaries/RedisDictionarySource.cpp

```diff
@@ -250,23 +222,64 @@ namespace DB
                 if (isInteger(type))
                     key << DB::toString(key_columns[i]->get64(row));
                 else if (isString(type))
-                    key << get<String>((*key_columns[i])[row]);
+                    key << get<const String &>((*key_columns[i])[row]);
                 else
                     throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected type of key in Redis dictionary");
             }
 
             keys.add(key);
         }
 
-        return Pipe(std::make_shared<RedisSource>(client, std::move(keys), storage_type, sample_block, max_block_size));
+        return Pipe(std::make_shared<RedisSource>(std::move(conn), std::move(keys), storage_type, sample_block, REDIS_MAX_BLOCK_SIZE));
     }
 
-
     String RedisDictionarySource::toString() const
     {
         return "Redis: " + host + ':' + DB::toString(port);
     }
 
+    auto RedisDictionarySource::getConnection() const -> ConnectionPtr
+    {
+        ClientPtr client;
+        bool ok = pool->tryBorrowObject(client,
+            [] { return std::make_unique<Poco::Redis::Client>(); },
```

Comment:
Can `std::make_unique<Poco::Redis::Client>()` throw? It looks like if it throws, the borrow object pool might get broken https://github.com/ClickHouse/ClickHouse/blob/master/base/base/BorrowedObjectPool.h#L132-L138. When I was adding this borrowed pool for postgres pool, Maxim suggested to return nullptr here and to initialize afterwards  https://github.com/ClickHouse/ClickHouse/blob/master/src/Core/PostgreSQL/PoolWithFailover.cpp#L75